File: algorithms/python/优美的排列/优美的排列.py
# Source : https://leetcode.cn/problems/beautiful-arrangement/description/
# Author : Yanan Wang
# Date   : 2023-09-06

# 先画一颗决策树       
# 剪枝条件有两个: 
#       1、使用过的数字不能进行递归
#       2、不能满足整除条件的不能进行递归

class Solution:
    def countArrangement(self, n: int) -> int:
        ans, used = 0, [False for _ in range(n + 1)]
        def dfs(index = 1):
            if n + 1 == index:
                nonlocal ans
                ans += 1
                return
            for i in range(1, n + 1):
                if not used[i] and (i % index == 0 or index % i == 0):
                    used[i] = True
                    dfs(index + 1)
                    used[i] = False

        dfs()
        return ans

File: algorithms/cpp/优美的排列/优美的排列.cpp
// Source : https://leetcode.cn/problems/beautiful-arrangement/description/
// Author : Yanan Wang
// Date   : 2023-09-07

// 本题采用递归回溯解法：
// 先画一颗决策树：根据下标[1, n]，然后1-n的数字往下标里填
//        需要剪枝的地方：
//                      1、 用过的数字不能再使用了
//                      2、 不能被整除的不能进行递归

class Solution {
private:
    int ans;
    vector<bool> used;
public:
    void dfs(int n, int index) {
        if (n + 1 == index) {
            ++ans;
            return;
        }
        for (int i = 1; i <= n; ++i) {
            if (!used[i] && (i % index == 0 || index % i == 0)) {
                used[i] = true;
                dfs(n, index + 1);
                used[i] = false;
            }
        }
    }
    int countArrangement(int n) {
        ans = 0;
        used = vector<bool>(n + 1, false);
        dfs(n, 1);
        return ans;
    }
};
